crates/sui-framework/packages/sui-system/sources/validator_set.move

```diff
@@ -233,11 +233,37 @@ module sui_system::validator_set {
             EDuplicateValidator
         );
         assert!(validator.is_preactive(), EValidatorNotCandidate);
-        assert!(validator.total_stake_amount() >= min_joining_stake_amount, EMinJoiningStakeNotReached);
+        assert!(can_join(self, validator.total_stake_amount(), ctx), EMinJoiningStakeNotReached);
 
         self.pending_active_validators.push_back(validator);
     }
 
+    /// Return `true` if a  candidate validator with `stake` will have sufficeint voting power to join the validator set
+    fun can_join(self: &ValidatorSet, stake: u64, ctx: &TxContext): bool {
+        let (min_joining_voting_power, _, _) = get_voting_power_thresholds(ctx);
+        // if the validator will have at least `min_joining_voting_power` after joining, they can join.
+        // this formula comes from SIP-39 TODO link once landed
+        let future_total_stake = self.total_stake + stake;
+        let future_validator_voting_power = voting_power::derive_voting_power(stake, future_total_stake);
+        future_validator_voting_power >= min_joining_voting_power
+    }
+
+    /// return (min, low, very low voting power) thresholds
+    fun get_voting_power_thresholds(ctx: &TxContext): (u64, u64, u64) {
+        // TODO: adjust as needed depending on when this will ship. and how will we deal with the devnet/testnet/mainnet differences?
```

Comment:
probably best to use protocol version for this maybe?

Reply:
I looked into this, but I think it actually doesn't work unless we assume that upgrades happen every 2 weeks and every upgrade is a new protocol version. And since we'd have to hardcode the start protocol version anyway, we might as well encode the start epoch and keep it simple.

The good thing is that `start_epoch` can be set as conservatively as we want--e.g., if we anticipated this framework upgrade shipping on 10/30, there'd be no problem with setting the `start_epoch` as 11/30

Reply:
I more meant that once this code lands and has been upgraded on a network we can immediately start using it the next epoch. Hard coding an epoch number likely will not work since we have different networks at different epochs

Reply:
IIUC the nice Manos suggestion below will fix the X-network issue